blocks/algorithms.py

```diff
@@ -0,0 +1,227 @@
+"""Training algorithms."""
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+
+import theano
+from theano import tensor
+
+from blocks.utils import update_instance
+from blocks.graph import ComputationGraph
+
+
+class TrainingAlgorithm(object):
+    """Base class for training algorithms.
+
+    A training algorithm object has a simple life-cycle.
+    First it is initialized by calling its :meth:`initialize` method.
+    At this stage, for instance, Theano functions can be compiled.
+    After that the :meth:`process_batch` method is repeatedly
+    called with a batch of training data as a parameter.
+
+    """
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def initialize(self):
+        """Initialize the training algorithm."""
+        pass
+
+    @abstractmethod
+    def process_batch(self, batch):
+        """Process a batch of training data.
+
+        Attributes
+        ----------
+        batch : dict
+            A dictionary of (source name, data) pairs.
+
+        """
+        pass
+
+
+class TheanoTrainingAlgorithm(TrainingAlgorithm):
+    """A training algorithm based on Theano functions.
+
+    Parameters
+    ----------
+    cost : Theano variable
+        The objective to be minimized.
+    params : list of Theano shared variables
+        The parameters to be tuned.
+
+    Attributes
+    ----------
+    updates : list of (shared variable, Theano expression) tuples
+        Updates to be done for every batch. It is required that the
+        updates are done using the old values of optimized parameters.
+    cost : Theano variable
+        The objective to be minimized.
+    params : list of Theano shared variables
+        The parameters to be tuned.
+
+    Notes
+    -----
+        Changing `updates` attribute or calling `add_updates` after
+        the `initialize` method is called will have no effect.
+
+    .. todo::
+
+       Ideally the parameters should be extracted from the `cost`
+       computation graph. But in order for this to work our graph
+       traversal routine should recurse into ScanOp's, which is currently
+       not the case.
```

Comment:
Hey,

I came up with another solution:

Make a Model base class, that would be subclassed by the user for each 
experiment.
Then the class would have methods as Model.get_params() and 
Model.get_grads() that by default scan the graph and call theano to get 
gradients.

Then: the common case is easy - the user just subclasses Model(). When a 
weird model comes in, the user has to override the functions that get 
paremeters and gradients. Then the main loop stays theano agnostic, and 
the trainer doesn't deal with parameter extraction and gradient evaluation.



On 1/15/2015 6:11 PM, Dmitry Bogdanov wrote:
>
> In blocks/algorithms.py 
> <https://github.com/bartvm/blocks/pull/92#discussion-diff-23025530>:
>
> > +    cost : Theano variable
> > +        The objective to be minimized.
> > +    params : list of Theano shared variables
> > +        The parameters to be tuned.
> > +
> > +    Notes
> > +    -----
> > +        Changing `updates` attribute or calling `add_updates` after
> > +        the `initialize` method is called will have no effect.
> > +
> > +    .. todo::
> > +
> > +       Ideally the parameters should be extracted from the `cost`
> > +       computation graph. But in order for this to work our graph
> > +       traversal routine should recurse into ScanOp's, which is currently
> > +       not the case.
>
> We can not be and we are not fully Theano agnostic. But neither should 
> we assume that everything is a Theano variable. That being said I 
> would not like to see the assumption that the cost is a Theano 
> variable in the main loop. For instance, that would forbid forever 
> non-differentiable costs based on sampling.
>
> I also like the idea that every part of the library should be useful 
> on its own. But this is exactly why I want the training algorithms 
> that take a differentiable Theano expression for cost as an input to 
> have a default way to extract the parameters from the computation 
> graph, and not rely on the main loop for this simple (assuming that 
> using the |ComputationGraph| is allowed) job.
>
> And just a side note: I did not mean |ComputationGraph| to be a model 
> (and moreover /the/) model. I perceive it as a tool to work with 
> annotated Theano graphs.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/bartvm/blocks/pull/92/files#r23025530>.
>